ignite/handlers/checkpoint.py

```diff
@@ -728,9 +730,12 @@ def __init__(
     def last_checkpoint(self) -> Union[str, None]:
         if len(self._saved) < 1:
             return None
-        return os.path.join(self.save_handler.dirname, self._saved[-1].filename)
+        if isinstance(self.save_handler, DiskSaver):
```

Comment:
In `ModelCheckpoint`, `self.save_handler` is `DiskSaver`

Reply:
Yes, exactly! But it seems `mypy` cannot infer the current class that, here is what I'm getting without `if isinstance(self.save_handler, DiskSaver)`:

```bash
error: Item "function" of
"Union[Callable[..., Any], BaseSaveHandler]" has no attribute "dirname" 
[union-attr]
            return os.path.join(self.save_handler.dirname, self._saved[-1]...
                                ^
ignite/handlers/checkpoint.py:734: error: Item "BaseSaveHandler" of
"Union[Callable[..., Any], BaseSaveHandler]" has no attribute "dirname" 
[union-attr]
            return os.path.join(self.save_handler.dirname, self._saved[-1]...
```

So, instead of doing `if isinstance(...)`.  we can solve this with `  # type: ignore`, anyway we are totally sure what is going on here, What do you think?

Reply:
@trsvchn and if we add type hint where we set `self.save_handler = DiskSaver(...)`, it does not help to mypy ? 
@gruebel may I ask what do you think what can be done here ?

Reply:
@vfdev-5 no type hinting `self.save_handler = DiskSaver(...)` do not help as well

Reply:
@trsvchn how about adding a code that raises an exception if `self.save_handler` is not `DiskSaver`. As `self.save_handler` is a public and user can redefine it between construction and `last_checkpoint` call...

Reply:
@vfdev-5 Yes! I think it's a great idea! Looks like `mypy` really helped :)
```suggestion
        if isinstance(self.save_handler, DiskSaver):
            return os.path.join(self.save_handler.dirname, self._saved[-1].filename)
        else:
            raise RuntimeError("Unable to save checkpoint, invalid save_handler.")
```

Does it look fine?

Reply:
Personally, i'd write it like 
```python
if not isinstance(self.save_handler, DiskSaver):
    raise RuntimeError("Unable to save checkpoint, save_handler should be DiskSaver, got {}.".format(type(self.save_handler)))

return os.path.join(self.save_handler.dirname, self._saved[-1].filename)
```

Reply:
Yes, it looks cleaner! Thanks for the tip!